carma_wm/include/carma_wm/SignalizedIntersectionManager.h

```diff
@@ -0,0 +1,127 @@
+#pragma once
+
+/*
+ * Copyright (C) 2020-2021 LEIDOS.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+#include <ros/ros.h>
+#include <boost/uuid/uuid.hpp>
+#include <boost/uuid/uuid_io.hpp>
+#include <boost/uuid/uuid_generators.hpp>
+#include <lanelet2_core/LaneletMap.h>
+#include <lanelet2_core/primitives/Point.h>
+#include <lanelet2_extension/regulatory_elements/StopRule.h>
+#include <lanelet2_extension/regulatory_elements/CarmaTrafficSignal.h>
+#include <lanelet2_core/primitives/LaneletOrArea.h>
+#include <lanelet2_extension/projection/local_frame_projector.h>
+#include <carma_wm/WorldModelUtils.h>
+#include <cav_msgs/MapData.h>
+#include <lanelet2_core/Forward.h>
+#include <lanelet2_extension/regulatory_elements/SignalizedIntersection.h>
+#include <lanelet2_core/geometry/LaneletMap.h>
+namespace carma_wm
+{
+
+struct LANE_DIRECTION
+{
+  static const uint8_t INGRESS = 1;
+  static const uint8_t EGRESS = 2;
+};
+
+using namespace lanelet::units::literals;
+
+/*! \brief This class manages and keeps track of all signalized intersections in the map.
+           All of the SPAT and MAP standard's lane ids to lanelet id mapping is recorded here.
+ */
+class SignalizedIntersectionManager
+{
+public:
+  SignalizedIntersectionManager(){}
+
+  /*! 
+  *  \brief Create relevant signalized intersection and carma traffic signals based on the MAP.msg and the lanelet_map
+  *  \param intersections to return
+  *  \param traffic_signals to return
+  *  \param map_msg MAP.msg that consists all static data portion of the intersection
+  *  \param map lanelet_map to check 
+  */
+  void createIntersectionFromMapMsg(std::vector<lanelet::SignalizedIntersectionPtr>& intersections, std::vector<lanelet::CarmaTrafficSignalPtr>& traffic_signals, const cav_msgs::MapData& map_msg, const std::shared_ptr<lanelet::LaneletMap>& map);
+
+  /*! 
+  *  \brief Returns mapping of MAP lane id to lanelet id for the given map and intersection.msg in the MAP.msg.
+            This function also records signal_group_id_to its own lanelet id, and also signal group to entry and exit lanelets id mappings
+  *  \param entry lane id to lanelet id mappings to return
+  *  \param exit  lane id to lanelet id mappings to return
+  *  \param intersection MAP.msg that consists all static data portion of the intersection
+  *  \param map lanelet_map to check 
+  *  \throw invalid_argument if given coordinates in the msg doesn't exist in the map
```

Comment:
Fine for now, but we might want to think about some kind of error buffer in the future. Map msgs are made from google maps or open streets maps normally so we might run into some errors from that. Hopefully proj will get us close enough

Reply:
added comment for it